### ISSUE TITLE
ImageNet Distributed Training with Distributed Validation

### ISSUE THREAD

**Issue by teng-li** [pytorch/examples#461]
This is on my TODO for a while. Put it here in case I forget.

**Comment by xvjiarui** [comment]
Looking forward to it. Since the current [example script](https://github.com/pytorch/examples/blob/master/imagenet/main.py) does validation for every process. If I am training on 4 nodes with 4 GPUs on each, there would be 16 validation processing running at the same time, which is not quite ideal.

**Comment by shubhvachher** [comment]
Hey all. Bumping this for a clarification.

Does this not entail just adding a distributed sampler, as for train in line 209, to `val_loader` at
https://github.com/pytorch/examples/blob/1de2ff9338bacaaffa123d03ce53d7522d5dcc2e/imagenet/main.py#L218
which, for now, as @xvjiarui points out would be blowing up every process with unnecessary work?

PS: Github issue search got me here when I was opening a new issue. Kudos to them!

**Comment by msalvaris** [comment]
Bumping again :)
Besides using the distributed sampler wouldn't we need to also aggregate the metrics from the GPUs with all_gather?

**Comment by soumith** [comment]
> I also have an example that runs on Apex and there I am able to simply run the validation on Rank 0 (simple if branching statement) but in normal PyTorch the same logic seems to lock up. Does someone know the reason for this difference?

the all_reduce step will still be run, and the rank=0 node will be waiting on rank > 0 nodes to get gradients. It's dumb but I think it's fixed on PyTorch master (with the param `find_unused_parameters=True` given to `nn.DistrbutedDataParallel`).
In your case, if you are running validation only on rank 0, then you should just run validation on `model.module(input)`, rather than `model(input)`, so that you bypass all the distributed logic

**Comment by msalvaris** [comment]
Fantastic, thanks @soumith

**Comment by leimao** [comment]
So far, I saw there has not been distributed validation implemented in this example. Is my understanding correct?

**Comment by hologerry** [comment]
Thank you so much! @soumith

**Comment by hkunzhe** [comment]
> Bumping again :)
> Besides using the distributed sampler wouldn't we need to also aggregate the metrics from the GPUs with all_gather?
> 
> I also have an example that runs on Apex and there I am able to simply run the validation on Rank 0 (simple if branching statement) but in normal PyTorch the same logic seems to lock up. Does someone know the reason for this difference?

In my env (Pytorch 1.5.1) , it does work and won't lock up.

**Comment by yucoian** [comment]
> This is on my TODO for a while. Put it here in case I forget.
 Just a reminder!
